packages/api-server/src/schema/squiggle.ts

```diff
@@ -0,0 +1,160 @@
+import { builder } from "../builder";
+import { prisma } from "../prisma";
+
+function getKey(code: string): string {
+  return crypto.createHash("md5").update(code).digest("base64");
+}
+
+const squiggleValueToJSON = (value: SqValue) => {
+  // this is a lazy shortcut to traverse the value tree; should be reimplemented without parse/stringify
+  return JSON.parse(
+    JSON.stringify(value.asJS(), (key, value) => {
+      if (value instanceof Map) {
+        return Object.fromEntries(value.entries());
+      }
+      if (value instanceof SqAbstractDistribution) {
+        return value.toString();
+      }
+      return value;
+    })
+  );
+};
+
+type SquiggleOutput = {
+  isCached: boolean;
+} & (
+  | {
+      isOk: false;
+      errorString: string;
+      resultJSON?: undefined;
+      bindingsJSON?: undefined;
+    }
+  | {
+      isOk: true;
+      errorString?: undefined;
+      resultJSON: Prisma.JsonValue;
+      bindingsJSON: Prisma.JsonValue;
+    }
+);
+
+function runSquiggle(code: string): SquiggleOutput {
+  const MAIN = "main";
+
+  const project = SqProject.create();
+
+  project.setSource(MAIN, code);
+  project.run(MAIN);
+
+  const result = project.getResult(MAIN);
+  const bindings = project.getBindings(MAIN);
+
+  return result.ok
+    ? {
+        isCached: false,
+        isOk: true,
+        resultJSON: squiggleValueToJSON(result.value),
+        bindingsJSON: squiggleValueToJSON(bindings.asValue()),
+      }
+    : {
+        isCached: false,
+        isOk: false,
+        errorString: result.value.toString(),
+      };
+}
+
+const SquiggleOutputObj = builder
+  .interfaceRef<SquiggleOutput>("SquiggleOutput")
+  .implement({
+    fields: (t) => ({
+      isCached: t.exposeBoolean("isCached"),
+    }),
+  });
+
+builder.objectType(
+  builder.objectRef<Extract<SquiggleOutput, { isOk: true }>>(
+    "SquiggleOkOutput"
+  ),
+  {
+    name: "SquiggleOkOutput",
+    interfaces: [SquiggleOutputObj],
+    isTypeOf: (value) => (value as SquiggleOutput).isOk,
+    fields: (t) => ({
+      resultJSON: t.string({
+        resolve(obj) {
+          return JSON.stringify(obj.resultJSON);
+        },
+      }),
+      bindingsJSON: t.string({
+        resolve(obj) {
+          return JSON.stringify(obj.bindingsJSON);
+        },
+      }),
+    }),
+  }
+);
+
+builder.objectType(
+  builder.objectRef<Extract<SquiggleOutput, { isOk: false }>>(
+    "SquiggleErrorOutput"
+  ),
+  {
+    name: "SquiggleErrorOutput",
+    interfaces: [SquiggleOutputObj],
+    isTypeOf: (value) => !(value as SquiggleOutput).isOk,
+    fields: (t) => ({
+      errorString: t.exposeString("errorString"),
+    }),
+  }
+);
+
+builder.queryField("runSquiggle", (t) =>
+  t.field({
+    type: SquiggleOutputObj,
+    args: {
+      code: t.arg.string({ required: true }),
+    },
+    async resolve(_, { code }) {
+      const key = getKey(code);
+
+      const cached = await prisma.squiggleCache.findUnique({
+        where: { id: key },
```

Comment:
Making the key the ID makes me uncomfortable. In the future, I imagine we might have multiple generations for the same code. Also, the client would ideally pass in the Sample Count, which would impact this.

Reply:
We could mix other values in the hashing function too. It would invalidate the existing cache in some cases, but it doesn't seem like a big problem.

Reply:
I didn't see the need to add another synthetic primary key id, since all lookups are going to happen by some hash of code and other parameters.

Reply:
Keys are supposed to be unique. If we save multiple runs for the same code/params, then how will that work? Or we just don't do that?

Reply:
We could have a separate run id. But then the client would need to remember the execution id, to know which version of the run to pick, so it could be mixed in the same key.

Reply:
Why would the client care which version to pick? All would be for the same inputs. I assume, as a client, I'd be happy to just have n options for the same one, then it could be up to me. As long as responses are deterministic, I could just use the First response, if I want consistency.

Reply:
>  it could be mixed in the same key.

My impression was that it's kinda bad to have two rows with the same Key? How could people keep them apart? I thought keys should be unique.

Reply:
Oh, I see. Yes, if you want to pick a run out of several older runs, on the client side, then we'd need a separate key for that.

But maybe that could be `code` or `codeHash` field that's not unique, and `key` could still be unique.